Introduce if/switch expressions

hamishknight opened the pull request:
Introduce SingleValueStmtExpr, which allows the embedding of a statement in an expression context. This then allows us to parse and type-check `if` and `switch` statements as expressions. In the future, SingleValueStmtExpr could also be used for e.g `do` expressions.

For now, only single expression branches are supported for producing a value from an `if`/`switch` expression, and each branch is type-checked independently. A multi-statement branch may only appear if it ends with a `throw`, and it may not `break`, `continue`, or `return`.

The placement of `if`/`switch` expressions is also currently limited by a syntactic use diagnostic. Currently they're only allowed in bindings, assignments, throws, and returns. But this could be lifted in the future if desired.

xedin committed:
[AST] Expand TypeJoin expression to support joining over a type

(cherry picked from commit 4efc35a76c7f1a6119c5ffc9fdaaa6766c895ba6)

xedin committed:
[CSClosure] Allocate temporary type-join and opaque value exprs in solver arena

(cherry picked from commit f3df7ab31b8a271fba4f6f3374e182b5154b2636)

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented on include/swift/Sema/ConstraintLocator.h:
```diff
@@ -85,6 +85,15 @@ enum ContextualTypePurpose : uint8_t {
+  CTP_SingleValueStmtBranch, ///< The contextual type for a branch in a single
+                             ///< value statement expression.
+
+  CTP_SingleValueStmtBranchInSingleExprClosure, ///< The contextual type for a
+                                                ///< branch in a single value
+                                                ///< statement expression that's
+                                                ///< part of a single expression
+                                                ///< return in a closure.
```
@xedin I don't love this, but it's needed to correctly handle the Void conversion for a single expression body closure. I don't suppose you can think of a better way of doing this? Ideally we'd just store a locator and get the information from there, but that might be a bit overkill.

xedin replied:
Have you considered this kind of a scheme:

a. allocate a single type variable that represents if/switch expression and just like with closures (and soon with taps) - provide a default type and delay until it either gets bound to contextual type or that default (default would just mean - proceed to step b.).

b. Once it's time to resolve, you can employ a trick similar to result builder transform and result type joining I implemented in (https://github.com/apple/swift/pull/62806):
  1. Tweak `TypeJoinExpr` to accept `SingleValueStmtExpr *` and generate a join based on branches (it can use `cs.getType` to extract type of each branch instead of walking into it).
  2. Use syntactic element generator to produce a conjunction and add `TypeJoinExpr` as the last element.
  3. Record conjunction as a new constraint.

This would mean that you don't have to allocate any extra variables, mess with brace elements and store contextual type information.

hamishknight commented on lib/Sema/CSBindings.cpp:
```diff
@@ -1071,6 +1071,15 @@ bool BindingSet::favoredOverConjunction(Constraint *conjunction) const {
       return false;
   }
+
+  if (conjunction->getLocator()->isForSingleValueStmtConjunction()) {
+    // If the bindings should be delayed, try the conjunction first.
+    // TODO: Seems like this ought to apply generally? Currently regresses a
+    // diagnostic in a test.
+    if (isDelayed())
+      return false;
+  }
```
@xedin Would this make sense to apply more generally? I haven't had a chance to dig into the diagnostic regression. The case it's handling for if/switch expressions is if they're in an array literal, this stops us eagerly binding `[Any]` without solving the conjunction. That being said, that's currently a case we don't allow, so I think we could totally rip this out for now.

xedin replied:
Do we still need this? I think this conjunction should behave just like a regular one?

hamishknight replied:
The need for it is still present, at least if we ever allow if/switch expressions in array literals. That being said, I'm fine with removing for now given that we don't yet allow them there (we'll just emit an, extra, spurious type error).

The reason this hasn't been needed before is that closure solving is triggered by type variable binding, not conjunction solving, so this isn't handled by the existing delaying logic.

xedin replied:
I'd say let's remove it and cross that bridge when we get there, it's possible that we might work out a more general solution before that point.

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

xedin commented:
I'll try to take a look at this today!

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -1187,6 +1205,53 @@ bool ConstraintSystem::generateConstraints(AnyFunctionRef fn, BraceStmt *body) {
   return generator.hadError;
 }
 
+bool ConstraintSystem::generateConstraints(SingleValueStmtExpr *E) {
```
We need to be able to access `SyntacticElementConstraintGenerator`, which is currently local to this file

xedin replied:
I see.

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

hamishknight commented:
Diff of changes modulo the rebase: https://github.com/apple/swift/compare/a6b33578a22e10de7ce272a60a1120f757f43f53..96da632832166b6d39b7f6b14f68560b50956210

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -2208,29 +2383,15 @@ bool ConstraintSystem::applySolutionToBody(Solution &solution,
 
-  fn.setTypecheckedBody(castToStmt<BraceStmt>(body),
-                        fn.hasSingleExpressionBody());
+  fn.setTypecheckedBody(cast<BraceStmt>(body), fn.hasSingleExpressionBody());
```
Hm, I think we have a bug here unrelated to your changes - if function/accessor/closure has been result builder transformed it would no longer be `single-expression` so we need to expand the check.

hamishknight replied:
Heh, good catch! I can file a bug if you want

xedin replied:
Let's do that so it doesn't get lost. I'll open a PR to fix that once your changes land.

hamishknight replied:
https://github.com/apple/swift/issues/63264

xedin commented on lib/Sema/CSSyntacticElement.cpp:
```diff
@@ -1195,6 +1247,87 @@ bool ConstraintSystem::generateConstraints(AnyFunctionRef fn, BraceStmt *body) {
+  // expression if the result is known to be Void.
+  if (auto *CE = dyn_cast<ClosureExpr>(E->getDeclContext())) {
+    if (CE->hasSingleExpressionBody() &&
+        CE->getSingleExpressionBody()->getSemanticsProvidingExpr() == E) {
```
We have `hasExplicitResult`, could it be used here?

hamishknight replied:
Nice, yup let's use that

xedin reviewed (approved):
LGTM! Left a couple of minor comments that shouldn't block landing of this PR.

hamishknight commented:
Likely still stuck on the swift-syntax issue, but:

https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
Likely still stuck on the swift-syntax issue, but:

https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test source compatibility

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please SourceKit stress test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please SourceKit stress test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please SourceKit stress test

hamishknight committed:
[CS] NFC: Factor out brace element handling in a couple of cases

Split out brace element handling for a couple of
walkers into a new function. The diff for this is
best viewed without whitespace changes. This
should be NFC.

hamishknight committed:
[Sema] Avoid forming double braced single expression closures

For a single expression closure, just use the
expression as the body in the case where we're
coercing to Void, as the return is already
implied. This avoids crashing in
`ClosureExpr::getSingleExpressionBody` with a
double braced body.

Surprisingly it seems nothing is currently calling
`ClosureExpr::getSingleExpressionBody` after
type-checking, so no test case, but later commits
in this patch will exercise this case.

hamishknight committed:
[CS] NFC: Move SolverSnapshot::applySolution into the cpp file

hamishknight committed:
[Sema] Introduce PreCheckReturnStmtRequest

Factor out some type-checking logic for ReturnStmt,
including the conversion to FailStmt, into a
request. We can then invoke this request from
both the regular type-checking path, as well as
during a pre-check of an expression.

hamishknight committed:
Introduce if/switch expressions

Introduce SingleValueStmtExpr, which allows the
embedding of a statement in an expression context.
This then allows us to parse and type-check `if`
and `switch` statements as expressions, gated
behind the `IfSwitchExpression` experimental
feature for now. In the future,
SingleValueStmtExpr could also be used for e.g
`do` expressions.

For now, only single expression branches are
supported for producing a value from an
`if`/`switch` expression, and each branch is
type-checked independently. A multi-statement
branch may only appear if it ends with a `throw`,
and it may not `break`, `continue`, or `return`.

The placement of `if`/`switch` expressions is also
currently limited by a syntactic use diagnostic.
Currently they're only allowed in bindings,
assignments, throws, and returns. But this could
be lifted in the future if desired.

hamishknight committed:
[Sema] Error on redundant try/awaits for if/switch expressions

Look through `try`/`await` markers when looking for
out of place if/switch expressions, and customize
the effect checking diagnostic such that we error
that `try`/`await` are redundant on `if`/`switch`
expressions.

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please SourceKit stress test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213
https://github.com/apple/swift-source-compat-suite/pull/764

@swift-ci please SourceKit stress test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please build toolchain

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test macOS

hamishknight commented:
https://github.com/apple/swift-syntax/pull/1228
https://github.com/apple/sourcekit-lsp/pull/672
https://github.com/apple/swift-format/pull/452
https://github.com/apple/swift-stress-tester/pull/213

@swift-ci please test Linux

hamishknight commented:
:ship:
